// Code generated by ent, DO NOT EDIT.

package ent

import (
	"context"
	"frog-go/internal/ent/predicate"
	"frog-go/internal/ent/teste"

	"entgo.io/ent/dialect/sql"
	"entgo.io/ent/dialect/sql/sqlgraph"
	"entgo.io/ent/schema/field"
)

// TesteDelete is the builder for deleting a Teste entity.
type TesteDelete struct {
	config
	hooks    []Hook
	mutation *TesteMutation
}

// Where appends a list predicates to the TesteDelete builder.
func (td *TesteDelete) Where(ps ...predicate.Teste) *TesteDelete {
	td.mutation.Where(ps...)
	return td
}

// Exec executes the deletion query and returns how many vertices were deleted.
func (td *TesteDelete) Exec(ctx context.Context) (int, error) {
	return withHooks(ctx, td.sqlExec, td.mutation, td.hooks)
}

// ExecX is like Exec, but panics if an error occurs.
func (td *TesteDelete) ExecX(ctx context.Context) int {
	n, err := td.Exec(ctx)
	if err != nil {
		panic(err)
	}
	return n
}

func (td *TesteDelete) sqlExec(ctx context.Context) (int, error) {
	_spec := sqlgraph.NewDeleteSpec(teste.Table, sqlgraph.NewFieldSpec(teste.FieldID, field.TypeUint))
	if ps := td.mutation.predicates; len(ps) > 0 {
		_spec.Predicate = func(selector *sql.Selector) {
			for i := range ps {
				ps[i](selector)
			}
		}
	}
	affected, err := sqlgraph.DeleteNodes(ctx, td.driver, _spec)
	if err != nil && sqlgraph.IsConstraintError(err) {
		err = &ConstraintError{msg: err.Error(), wrap: err}
	}
	td.mutation.done = true
	return affected, err
}

// TesteDeleteOne is the builder for deleting a single Teste entity.
type TesteDeleteOne struct {
	td *TesteDelete
}

// Where appends a list predicates to the TesteDelete builder.
func (tdo *TesteDeleteOne) Where(ps ...predicate.Teste) *TesteDeleteOne {
	tdo.td.mutation.Where(ps...)
	return tdo
}

// Exec executes the deletion query.
func (tdo *TesteDeleteOne) Exec(ctx context.Context) error {
	n, err := tdo.td.Exec(ctx)
	switch {
	case err != nil:
		return err
	case n == 0:
		return &NotFoundError{teste.Label}
	default:
		return nil
	}
}

// ExecX is like Exec, but panics if an error occurs.
func (tdo *TesteDeleteOne) ExecX(ctx context.Context) {
	if err := tdo.Exec(ctx); err != nil {
		panic(err)
	}
}
